plugins/rule_engines/irods_rule_language/src/rules.cpp

```diff
@@ -61,6 +61,39 @@ int readRuleSetFromLocalFile( const char *ruleBaseName, const char *rulesFileNam
 
     return 0;
 }
+int readRuleSetFromBuffer(const char* ruleBaseName,
+                          char* ruleBody,
+                          RuleSet* ruleSet,
+                          Env* funcDesc,
+                          int* errloc,
+                          rError_t* errmsg,
+                          Region* r)
+{
+    FILE* file = fmemopen(ruleBody, strlen(ruleBody), "r");
+    char errbuf[ERR_MSG_LEN];
+    if (file == nullptr) {
+        snprintf(errbuf, ERR_MSG_LEN, "readRuleSetFromBuffer() could not read rule from buffer");
+        addRErrorMsg(errmsg, RULES_FILE_READ_ERROR, errbuf);
```

Comment:
Why do we use `addRErrorMsg` instead of the logger here?

Reply:
From what I can tell, it looks to be sort of an old fashioned "stack trace," where the call starting from `readRuleStructAndRuleSetFromBuffer()` down to this function will append messages to the list, and print them out using `rodsLog` if an error occurs. The stack doesn't seem _too_ deep, so I could replace them one-by-one; or, we could keep this and I could just replace the rodsLog at the top. Thoughts?

Reply:
It is a strange usage of an rError stack, but it seems to be correct here. The messages are logged in the calling function, like @FifthPotato said. I'm not sure how I feel about it, but since we are just copying an existing function that is tried and true, maybe we just let this ride for now.

Reply:
if we're not sure, and want to keep moving - consider a new issue pointing to this conversation... and we can unravel / discuss there.

Reply:
> It is a strange usage of an rError stack, but it seems to be correct here. The messages are logged in the calling function, like @FifthPotato said. I'm not sure how I feel about it, but since we are just copying an existing function that is tried and true, maybe we just let this ride for now.

I'm ambivalent. On one hand, it seems to only extend 1-2 levels deep at most, meaning replacing them all shouldn't be too bad. It also saves a `malloc` and `free`, which counts as modernization. On the other hand, it is tried-and-true. What's the overall risk of replacing these here?

Reply:
I think the main risk would be a change in behavior when running rules. The error information returned would need to be done *somehow* to be a valid replacement for the `rError` stack, and even then, the output may differ. I'm not sure if our tests depend on that behavior or not, but that would be the first place we would see it (if there is indeed a change in behavior).

I'm thinking that this implementation should be put into the upcoming release and we make a new issue, like @trel said. That being said, feel free to play around with it and see if the behavior changes.

Reply:
If we're unsure of that line, then let's do what @trel said and keep it moving.